api/v1/verticadb_webhook.go

```diff
@@ -1701,6 +1728,204 @@ func (v *VerticaDB) checkSandboxPrimary(allErrs field.ErrorList, oldObj *Vertica
+	newSandboxMap := v.GenSandboxMap()
+	newSandboxIndexMap := v.GenSandboxIndexMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	oldSubclusterIndexMap := oldObj.GenSubclusterIndexMap()
+
+	for _, sandboxName := range persistSandboxes {
+		oldSandbox := oldSandboxMap[sandboxName]
+		newSandbox := newSandboxMap[sandboxName]
+		if oldSandbox.Image != newSandbox.Image {
+			newSandboxIndex := newSandboxIndexMap[sandboxName]
+			errMsg := v.checkSboxForShutdown(newSandbox, newSubclusterMap, oldSubclusterIndexMap)
+			if errMsg != "" {
+				p := field.NewPath("spec").Child("sandboxes").Index(newSandboxIndex)
+				err := field.Invalid(p.Child("image"),
+					sandboxName,
+					fmt.Sprintf("cannot change the image for sandbox %q because %q",
+						sandboxName, errMsg))
+				allErrs = append(allErrs, err)
+			}
+		}
+	}
+	return allErrs
+}
+
+// ensure the sandbox to terminate is not shut down
+func (v *VerticaDB) validateTerminatingSandboxes(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSandboxMap := v.GenSandboxMap()
+	sandboxesToBeRemoved := vutil.MapKeyDiff(oldSandboxMap, newSandboxMap)
+	newSubclusterMap := v.GenSubclusterMap()
+	for _, sandboxToBeRemoved := range sandboxesToBeRemoved {
+		numOfSclustersInSbox := len(oldSandboxMap[sandboxToBeRemoved].Subclusters)
+		sclustersToTerminate := numOfSclustersInSbox
+		for _, subcluster := range oldSandboxMap[sandboxToBeRemoved].Subclusters {
+			if _, ok := newSubclusterMap[subcluster.Name]; ok {
+				sclustersToTerminate--
+			}
+		}
+		// sandboxToBeRemoved and all its subclusters are not found in new spec
+		if sclustersToTerminate == numOfSclustersInSbox {
+			if oldSandboxMap[sandboxToBeRemoved].Shutdown {
+				p := field.NewPath("spec").Child("sandboxes")
+				err := field.Invalid(p,
+					sandboxToBeRemoved,
+					fmt.Sprintf("cannot terminate sandbox %q that has Shutdown field set to true",
+						sandboxToBeRemoved))
+				allErrs = append(allErrs, err)
+			} // else looks good
+		} // else is expected
+	}
+	return allErrs
+}
+
+/* func (v *VerticaDB) checkSboxForShutdownInSpecAndStatus(sandbox *Sandbox, subclusterMap map[string]*Subcluster, subclusterIndexMap map[string]int) string {
+	errMsgs = v.checkSboxForShutdownInSpec(sandbox, subclusterMap)
+	for _, subcluster := range sandbox.Subclusters {
+		if v.Status.Subclusters[subclusterIndexMap[subcluster.Name]].Shutdown {
+			errMsgs = append(errMsgs, fmt.Sprintf("Shutdown status of subcluster %q is true", subcluster.Name))
+		}
+	}
+	return errMsgs
+} */
+
+func (v *VerticaDB) checkSboxForShutdown(sandbox *Sandbox, newSClusterMap map[string]*Subcluster, oldSClusterIndexMap map[string]int) string {
+	if sandbox.Shutdown {
+		return fmt.Sprintf("Shutdown field of sandbox %q is set to true", sandbox.Name)
+	}
+	for _, subcluster := range sandbox.Subclusters {
+		i, foundInOldSBox := oldSClusterIndexMap[subcluster.Name]
```

Comment:
Why are you getting the index from old subcluster to use it in new? 
Maybe add a function to get a map[string]int to get the subcluster index in status. Most of the time the index is spec is the same in status but there are some exceptions we do not want to deal with. Anyway, I think the index should be from new vdb. We don't care if the subcluster was in old vdb sandbox

Reply:
Let me know if you have objections.

Reply:
That is a good point. I will make the change.